identity-apps-core/apps/recovery-portal/src/main/webapp/password-recovery-otp.jsp

```diff
@@ -0,0 +1,308 @@
+        request.setAttribute("errorMsg", errorMsg);
+        request.getRequestDispatcher("error.jsp").forward(request, response);
+    }
+%>
+<%
+    final String USERNAME_CLAIM_URI_FOR_TENANT = "http://wso2.org/claims/username";
+    final RecoveryApiV2 recoveryApiV2 = new RecoveryApiV2();
+    final String username = IdentityManagementEndpointUtil.getStringValue(request.getAttribute("username"));
+
+    String recoveryStage = Encode.forHtmlAttribute(request.getParameter("recoveryStage"));
+
+    if (RecoveryStage.INITIATE.equalsValue(recoveryStage)) {
+        // if otp is supported by a new channel (eg: email) update this value assignment. null means unsupported.
+        final String targetChannel = IdentityManagementEndpointConstants.PasswordRecoveryOptions.SMSOTP
+            .equals((String)request.getAttribute("channel"))
+                ? "SMS"
+                : null;
+        // Manage unsupported channel
+        if (StringUtils.isBlank(targetChannel)) {
+            redirectToErrorPageWithMessage(request, response, "Unknown.channel");
+            return;
+        }
+        List<UserClaim> userClaims = new ArrayList<UserClaim>();
+
+        // Get the username claim string for the tenant
+        UserClaim userNameClaim = new UserClaim();
+        userNameClaim.setUri(USERNAME_CLAIM_URI_FOR_TENANT);
+        userNameClaim.setValue(MultitenantUtils.getTenantAwareUsername(username));
+        userClaims.add(userNameClaim);
+
+        String flawConfirmationCode = "";
+        String recoveryCode = "";
+        String channelId = "";
+        String screenValue = "";
+
+        try {
+            // Initiate password recovery.
+            RecoveryInitRequest recoveryInitRequest = new RecoveryInitRequest();
+            recoveryInitRequest.setClaims(userClaims);
+            Map<String, String> requestHeaders = new HashedMap();
+            if (request.getParameter("g-recaptcha-response") != null) {
+                requestHeaders.put("g-recaptcha-response", request.getParameter("g-recaptcha-response"));
+            }
+            List<AccountRecoveryType> resp = 
+                recoveryApiV2.initiatePasswordRecovery(recoveryInitRequest, tenantDomain, requestHeaders);
+            if (resp == null) {
+                /** Handle invalid username scenario. proceeds to next level without warning to 
+                avoid an attacker bruteforcing to learn the usernames. */
+                request.setAttribute("screenValue", "******" + getRandomNumberString(4, username));
+                request.setAttribute("resendCode", UUID.randomUUID().toString());
+                request.setAttribute("flowConfirmationCode", UUID.randomUUID().toString());
+                request.getRequestDispatcher("sms-otp.jsp").forward(request, response);
+                return;
+            }
+            for(AccountRecoveryType recoveryType: resp) {
+                if (recoveryType.getMode().equals("recoverWithNotifications")) {
```

Comment:
This can result in NPE. Let's refactor as follows. WDYT?

```suggestion
                if ("recoverWithNotifications".equals(recoveryType.getMode())) {
```

Reply:
addressed with
https://github.com/wso2/identity-apps/pull/5817/commits/877baa435b1c4ba3cfffef80cd1e5a56dc36d0c6